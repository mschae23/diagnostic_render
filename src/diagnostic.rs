//! Diagnostic data structures.
//!
//! End users are encouraged to create their own implementations
//! for their specific use cases, and convert them to this crate's
//! representation when needed.

use std::fmt::{Display, Formatter};
use std::ops::Range;

/// A severity level for diagnostic messages.
///
/// These are ordered in the following way:
///
/// ```rust
/// use diagnostic_render::diagnostic::Severity;
///
/// assert!(Severity::Bug > Severity::Error);
/// assert!(Severity::Error > Severity::Warning);
/// assert!(Severity::Warning > Severity::Note);
/// assert!(Severity::Note > Severity::Help);
/// ```
#[derive(Clone, Copy, Hash, Debug, PartialEq, Eq, PartialOrd, Ord)]
pub enum Severity {
    /// A help message
    Help,
    /// A note
    Note,
    /// A warning
    Warning,
    /// An error
    Error,
    /// An unexpected bug
    Bug,
}

impl Display for Severity {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", match self {
            Severity::Bug => "bug",
            Severity::Error => "error",
            Severity::Warning => "warning",
            Severity::Note => "note",
            Severity::Help => "help",
        })
    }
}

/// A style for annotations.
///
/// These are ordered in the following way:
/// ```rust
/// use diagnostic_render::diagnostic::AnnotationStyle;
///
/// assert!(AnnotationStyle::Primary < AnnotationStyle::Secondary);
/// ```
#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
pub enum AnnotationStyle {
    /// Annotations that describe the primary cause of a diagnostic.
    Primary,
    /// Annotations that provide additional context for a diagnostic.
    Secondary,
}

/// An annotation describing an underlined region of code associated with a diagnostic.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Annotation<FileId> {
    /// The style of the annotation.
    pub style: AnnotationStyle,
    /// The file that we are annotating.
    pub file_id: FileId,
    /// The range in bytes we are going to include in the final snippet.
    pub range: Range<usize>,
    /// An optional label to provide some additional information for the
    /// underlined code. These should not include line breaks.
    pub label: String,
}

impl<FileId> Annotation<FileId> {
    /// Create a new annotation with no label.
    pub fn new<R: Into<Range<usize>>>(style: AnnotationStyle, file_id: FileId, range: R) -> Self {
        Annotation {
            style,
            file_id,
            range: range.into(),
            label: String::new(),
        }
    }

    /// Create a new annotation with a style of [`AnnotationStyle::Primary`].
    ///
    /// [`AnnotationStyle::Primary`]: AnnotationStyle::Primary
    pub fn primary<R: Into<Range<usize>>>(file_id: FileId, range: R) -> Self {
        Self::new(AnnotationStyle::Primary, file_id, range)
    }

    /// Create a new label with a style of [`AnnotationStyle::Secondary`].
    ///
    /// [`AnnotationStyle::Secondary`]: AnnotationStyle::Secondary
    pub fn secondary<R: Into<Range<usize>>>(file_id: FileId, range: R) -> Self {
        Self::new(AnnotationStyle::Secondary, file_id, range)
    }

    /// Add a label to the annotation.
    pub fn with_label<L: ToString>(mut self, label: L) -> Self {
        self.label = label.to_string();
        self
    }
}

/// A note associated with the primary cause of a diagnostic.
/// They can be used to explain the diagnostic, or include help
/// on how to fix an issue.
///
/// They are displayed at the end of diagnostics, after the source code with
/// its annotations.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Note {
    /// The severity of the note.
    ///
    /// This should usually only be [`Severity::Help`] or [`Severity::Note`].
    ///
    /// [`Severity::Help`]: Severity::Help
    /// [`Severity::Note`]: Severity::Note
    pub severity: Severity,
    /// The message of this note.
    /// This can include line breaks for improved formatting.
    /// It should not be empty.
    pub message: String,
}

impl Note {
    /// Create a new note.
    pub fn new<M: ToString>(severity: Severity, message: M) -> Self {
        Note {
            severity, message: message.to_string(),
        }
    }

    /// Create a new note with a severity of [`Severity::Note`].
    ///
    /// [`Severity::Note`]: Severity::Note
    pub fn note<M: ToString>(message: M) -> Self {
        Self::new(Severity::Note, message)
    }

    /// Create a new note with a severity of [`Severity::Help`].
    ///
    /// [`Severity::Help`]: Severity::Help
    pub fn help<M: ToString>(message: M) -> Self {
        Self::new(Severity::Help, message)
    }
}

/// Represents a diagnostic message that can provide information like errors and
/// warnings to the user.
///
/// The position of a Diagnostic is considered to be the position of the [`Annotation`]
/// that has the earliest starting position and has the highest style which appears
/// in all the annotations of the diagnostic.
///
/// [`Annotation`]: Annotation
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Diagnostic<FileId> {
    /// The overall severity of the diagnostic.
    pub severity: Severity,
    /// An optional name or code that identifies this diagnostic.
    pub name: Option<String>,
    /// The main message associated with this diagnostic.
    ///
    /// These should not include line breaks, and in order support the 'short'
    /// diagnostic display style, the message should be specific enough to make
    /// sense on its own, without additional context provided by annotations and notes.
    pub message: String,
    /// Source annotations that describe the cause of the diagnostic.
    ///
    /// The order of the annotations inside the vector does not have any meaning.
    /// The annotations are always arranged in the order they appear in the source code.
    pub annotations: Vec<Annotation<FileId>>,
    /// Notes that are associated with the primary cause of the diagnostic.
    pub notes: Vec<Note>,

    // /// Additional diagnostics that can be used to show context from other files,
    // /// provide help by showing changed code, or similar. They are shown below notes.
    // pub sub_diagnostics: Vec<Diagnostic<FileId>>,

    /// The number of diagnostics following this one that are hidden due to
    /// something like panic mode in error reporting.
    pub suppressed_count: u32,
}

impl<FileId> Diagnostic<FileId> {
    /// Create a new diagnostic.
    pub fn new(severity: Severity) -> Self {
        Diagnostic {
            severity,
            name: None,
            message: String::new(),
            annotations: Vec::new(),
            notes: Vec::new(),
            suppressed_count: 0,
        }
    }

    /// Create a new diagnostic with a severity of [`Severity::Bug`].
    ///
    /// [`Severity::Bug`]: Severity::Bug
    pub fn bug() -> Self {
        Self::new(Severity::Bug)
    }

    /// Create a new diagnostic with a severity of [`Severity::Error`].
    ///
    /// [`Severity::Error`]: Severity::Error
    pub fn error() -> Self {
        Self::new(Severity::Error)
    }

    /// Create a new diagnostic with a severity of [`Severity::Warning`].
    ///
    /// [`Severity::Warning`]: Severity::Warning
    pub fn warning() -> Self {
        Self::new(Severity::Warning)
    }

    /// Create a new diagnostic with a severity of [`Severity::Note`].
    ///
    /// [`Severity::Note`]: Severity::Note
    pub fn note() -> Self {
        Self::new(Severity::Note)
    }

    /// Create a new diagnostic with a severity of [`Severity::Help`].
    ///
    /// [`Severity::Help`]: Severity::Help
    pub fn help() -> Self {
        Self::new(Severity::Help)
    }

    /// Set the name or code of the diagnostic.
    pub fn with_name<M: ToString>(mut self, name: M) -> Self {
        self.name = Some(name.to_string());
        self
    }

    /// Set the message of the diagnostic.
    pub fn with_message<M: ToString>(mut self, message: M) -> Self {
        self.message = message.to_string();
        self
    }

    /// Add an annotation to the diagnostic.
    pub fn with_annotation(mut self, annotation: Annotation<FileId>) -> Self {
        self.annotations.push(annotation);
        self
    }

    /// Add some labels to the diagnostic.
    pub fn with_annotations(mut self, mut annotations: Vec<Annotation<FileId>>) -> Self {
        self.annotations.append(&mut annotations);
        self
    }

    /// Add a note to the diagnostic.
    pub fn with_note(mut self, note: Note) -> Self {
        self.notes.push(note);
        self
    }

    /// Add some notes to the diagnostic.
    pub fn with_notes(mut self, mut notes: Vec<Note>) -> Self {
        self.notes.append(&mut notes);
        self
    }

    /// Sets the number of suppressed diagnostics.
    pub fn with_suppressed_count(mut self, suppressed_count: u32) -> Self {
        self.suppressed_count = suppressed_count;
        self
    }
}
