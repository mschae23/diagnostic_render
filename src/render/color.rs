use termcolor::{Color, ColorSpec, WriteColor};
use crate::diagnostic::{AnnotationStyle, Severity};

type Result = std::io::Result<()>;

/// Provides the terminal colors used in diagnostics.
///
/// Two default implementations are provided:
/// - [`DefaultColorConfig`], which should be similar to the colors used by `rustc`
/// - [`DisabledColorConfig`], which outputs no colors.
///
/// [`DefaultColorConfig`]: DefaultColorConfig
/// [`DisabledColorConfig`]: DisabledColorConfig
pub trait ColorConfig {
    /// Resets all style and formatting.
    fn reset(&self, f: &mut impl WriteColor) -> Result;

    /// Sets the formatting for annotations with a specific [`Severity`].
    ///
    /// [`Severity`]: Severity
    fn severity(&self, f: &mut impl WriteColor, severity: Severity) -> Result;

    /// Sets the formatting for annotations with [`Severity::Bug`].
    ///
    /// [`Severity::Bug`]: Severity::Bug
    fn bug(&self, f: &mut impl WriteColor) -> Result {
        self.severity(f, Severity::Bug)
    }

    /// Sets the formatting for annotations with [`Severity::Error`].
    ///
    /// [`Severity::Error`]: Severity::Error
    fn error(&self, f: &mut impl WriteColor) -> Result {
        self.severity(f, Severity::Error)
    }

    /// Sets the formatting for annotations with [`Severity::Warning`].
    ///
    /// [`Severity::Warning`]: Severity::Warning
    fn warning(&self, f: &mut impl WriteColor) -> Result {
        self.severity(f, Severity::Warning)
    }

    /// Sets the formatting for annotations with [`Severity::Note`].
    ///
    /// [`Severity::Note`]: Severity::Note
    fn note(&self, f: &mut impl WriteColor) -> Result {
        self.severity(f, Severity::Note)
    }

    /// Sets the formatting for annotations with [`Severity::Help`].
    ///
    /// [`Severity::Help`]: Severity::Help
    fn help(&self, f: &mut impl WriteColor) -> Result {
        self.severity(f, Severity::Help)
    }

    /// Sets the formatting for the optional error name or code.
    fn name(&self, f: &mut impl WriteColor, severity: Severity) -> Result;

    /// Sets the formatting for the main message of a diagnostic.
    fn message(&self, f: &mut impl WriteColor) -> Result;

    /// Sets the formatting for the file path, line and column numbers printed
    /// at the start of a code block.
    fn path(&self, f: &mut impl WriteColor) -> Result;

    /// Sets the formatting for the line number of a line of source code.
    fn line_number(&self, f: &mut impl WriteColor) -> Result;

    /// Sets the formatting for the separator between the line number and the line of source code.
    /// In most cases, this is either `" | "`, `"-->"`, or `"..."`.
    fn line_number_separator(&self, f: &mut impl WriteColor) -> Result;

    /// Sets the formatting for an annotation.
    /// The annotation style (primary or secondary) and the diagnostic severity are
    /// provided as context.
    ///
    /// The default configuration would redirect to [`Self::severity`] in the case of
    /// a primary annotation style, and use a specific formatting for the secondary
    /// annotation style.
    ///
    /// [`Self::severity`]: Self::severity
    fn annotation(&self, f: &mut impl WriteColor, style: AnnotationStyle, severity: Severity) -> Result;

    /// Sets the formatting for a line of source code.
    fn source(&self, f: &mut impl WriteColor) -> Result;

    fn note_severity(&self, f: &mut impl WriteColor, severity: Severity) -> Result;

    fn note_message(&self, f: &mut impl WriteColor, severity: Severity) -> Result;
}

/// The default color configuration.
/// This should be similar to the colors used in `rustc` diagnostics.
pub struct DefaultColorConfig;

impl ColorConfig for DefaultColorConfig {
    fn reset(&self, f: &mut impl WriteColor) -> Result {
        f.reset().into()
    }

    fn severity(&self, f: &mut impl WriteColor, severity: Severity) -> Result {
        f.set_color(ColorSpec::new().set_fg(Some(match severity {
            Severity::Help => Color::Green,
            Severity::Note => Color::Blue,
            Severity::Warning => Color::Yellow,
            Severity::Error | Severity::Bug => Color::Red,
        })).set_intense(matches!(severity, Severity::Note)).set_bold(matches!(severity, Severity::Note | Severity::Error | Severity::Bug)))
    }

    fn name(&self, f: &mut impl WriteColor, severity: Severity) -> Result {
        self.severity(f, severity)
    }

    fn message(&self, f: &mut impl WriteColor) -> Result {
        f.set_color(ColorSpec::new().set_bold(true))
    }

    fn path(&self, f: &mut impl WriteColor) -> Result {
        self.reset(f)
    }

    fn line_number(&self, f: &mut impl WriteColor) -> Result {
        f.set_color(ColorSpec::new().set_fg(Some(Color::Blue)).set_intense(true).set_bold(true))
    }

    fn line_number_separator(&self, f: &mut impl WriteColor) -> Result {
        f.set_color(ColorSpec::new().set_fg(Some(Color::Blue)).set_intense(true).set_bold(true))
    }

    fn annotation(&self, f: &mut impl WriteColor, style: AnnotationStyle, severity: Severity) -> Result {
        match style {
            AnnotationStyle::Primary => self.severity(f, severity),
            AnnotationStyle::Secondary => f.set_color(ColorSpec::new().set_fg(Some(Color::Blue)).set_intense(true).set_bold(true)),
        }
    }

    fn source(&self, f: &mut impl WriteColor) -> Result {
        self.reset(f)
    }

    fn note_severity(&self, f: &mut impl WriteColor, _severity: Severity) -> Result {
        f.set_color(ColorSpec::new().set_bold(true))
    }

    fn note_message(&self, f: &mut impl WriteColor, _severity: Severity) -> Result {
        self.reset(f)
    }
}

/// A no-op color configuration.
/// Sets no formatting and outputs no formatting codes.
pub struct DisabledColorConfig;

impl ColorConfig for DisabledColorConfig {
    fn reset(&self, _f: &mut impl WriteColor) -> Result {
        Ok(())
    }

    fn severity(&self, _f: &mut impl WriteColor, _severity: Severity) -> Result {
        Ok(())
    }

    fn name(&self, _f: &mut impl WriteColor, _severity: Severity) -> Result {
        Ok(())
    }

    fn message(&self, _f: &mut impl WriteColor) -> Result {
        Ok(())
    }

    fn path(&self, _f: &mut impl WriteColor) -> Result {
        Ok(())
    }

    fn line_number(&self, _f: &mut impl WriteColor) -> Result {
        Ok(())
    }

    fn line_number_separator(&self, _f: &mut impl WriteColor) -> Result {
        Ok(())
    }

    fn annotation(&self, _f: &mut impl WriteColor, _style: AnnotationStyle, _severity: Severity) -> Result {
        Ok(())
    }

    fn source(&self, _f: &mut impl WriteColor) -> Result {
        Ok(())
    }

    fn note_severity(&self, _f: &mut impl WriteColor, _severity: Severity) -> Result {
        Ok(())
    }

    fn note_message(&self, _f: &mut impl WriteColor, _severity: Severity) -> Result {
        Ok(())
    }
}
