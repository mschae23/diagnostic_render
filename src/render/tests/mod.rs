use termcolor::Buffer;
use crate::diagnostic::Severity;
use crate::file::SimpleFile;
use crate::render::color::DefaultColorConfig;
use super::*;

#[test]
fn test_header_1() {
    let mut buf = Buffer::no_color();
    let mut renderer = DiagnosticRenderer::new(&mut buf, DefaultColorConfig,
        SimpleFile::new("main.test", "unused source"),
        RenderConfig { surrounding_lines: 0 });
    renderer.render(vec![
        Diagnostic::new(Severity::Error)
            .with_name("test/diagnostic_1")
            .with_message("Test message")
    ]).unwrap();

    let buf = buf.into_inner();
    let result = String::from_utf8_lossy(&buf);

    insta::assert_snapshot!(result, @r###"
    error[test/diagnostic_1]: Test message
    "###);
}

/* // TODO this completely breaks the renderer right now
#[test]
fn test_fibonacci() {
    let source = r#"
    pub fn fibonacci(n: i32) -> u64 {
        if n < 0 {
            panic!("{} is negative!", n);
        } else if n == 0 {
            panic!("zero is not a right argument to fibonacci()!");
        } else if n == 1 {
            return 1;
        }

        let mut sum = 0;
        let mut last = 0;
        let mut curr = 1;
        for _i in 1..n {
            sum = last + curr;
            last = curr;
            curr = sum;
        }
        sum
    }
    "#;

    let mut buf = Buffer::no_color();
    let file = SimpleFile::new("test_file.test", source);

    let mut diagnostic = Diagnostic::new(Severity::Note)
        .with_message("A fibonacci function");

    {
        let mut opened = Vec::new();

        for (i, b) in source.bytes().enumerate() {
            let c = char::from(b);

            match c {
                '(' | '[' | '{' => opened.push((i, c)),
                '"' if opened.is_empty() || opened.last().unwrap().1 != '"' => {
                    opened.push((i, c));
                },
                ')' | ']' | '}' | '"' => {
                    if let Some((start, expected)) = opened.pop() {
                        if c == expected {
                            let range = start..i;
                            let label = match c {
                                ')' => "this is a pair of parenthesis",
                                ']' => "this is a pair of brackets",
                                '}' => "this is a pair of braces",
                                _ => "this is a string",
                            };
                            diagnostic = diagnostic.with_annotation(Annotation::new(AnnotationStyle::Secondary, (), range)
                                    .with_label(label));
                        }
                    }
                },
                _ => {},
            }
        }

        diagnostic = diagnostic.with_annotation(Annotation::new(AnnotationStyle::Primary, (), 0..source.len())
            .with_label("this is the whole program"));
    }

    let mut renderer = DiagnosticRenderer::new(&mut buf, DefaultColorConfig,
        file, RenderConfig { surrounding_lines: 0 });
    renderer.render(vec![diagnostic]).unwrap();

    let buf = buf.into_inner();
    let result = String::from_utf8_lossy(&buf);

    insta::assert_snapshot!(result);
}
*/

mod singleline;
mod ending;
mod starting;
