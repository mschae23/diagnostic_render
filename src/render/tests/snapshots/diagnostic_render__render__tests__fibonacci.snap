---
source: src/render/tests/mod.rs
expression: result
---
note: A fibonacci function
  --> test_file.test:1:1
 1 |       pub fn fibonacci(n: i32) -> u64 {
   |       ^               -------         -
   |  _____|               |               |
   | |  ___________________|_______________|
   | | |                   |
   | | |                   this is a pair of parenthesis
 2 | | |      if n < 0 {
   | | |  ______________-
 3 | | |          panic!("{} is negative!", n);
   | | |                 ---------------------
   | | |                 |||
   | | |                 ||this is a pair of braces
   | | |                 |this is a string
   | | |                 this is a pair of parenthesis
 4 | | |      } else if n == 0 {
   | | |______-                 -
   | |  ______|_________________|
   | | |      |
   | | |      this is a pair of braces
 5 | |            panic!("zero is not a right argument to fibonacci()!");
   | |                   -----------------------------------------------
   | |                   ||                                         |
   | |                   ||                                         this is a pair of parenthesis
   | |                   |this is a string
   | |                   this is a pair of parenthesis
 6 | |        } else if n == 1 {
   | |________-                 -
   |  ________|_________________|
   | |        |
   | |        this is a pair of braces
  ...|   
 8 | |        }
   | |________- this is a pair of braces
  ...     
13 |           for _i in 1..n {
   |  ________________________-
  ...|   
17 | |        }
   | |________- this is a pair of braces

